### DISCUSSION TITLE
Comparison Warning Mode, Group Image Editing

### DISCUSSION THREAD

**Discussion by Aiosa** [FRSOURCE/cypress-plugin-visual-regression-diff#277] [Ideas]
I wanted to use a warning mode that would let the test continue but show me the comparison issue nevertheless, but I could not find anything similar in the features. Also, I would like to delay comparing images to the end of testing and browse + accept/reject updates later. Then, re-run with error-like behaviour.

**What does it do?**
It issues an error/warning when the regression fails, does not fail the test, and offers later actions to perform (like full-replacement of the 'template' images).

**Why do I need it?**
Because Regression testing can break a lot, especially when the interface is modular and different deployments differ. I want the ability to compare such an interface against the 'original image' and consider whether changes are to be updated/accepted or whether the difference matters. I don't want to hide some UI elements since then the image does not capture the state well, elements can be hidden as a part of some other functionality, for example, so I wouldn't know whether the issue happens because the user has hidden the menu, since I would hide it every time. 
It also makes the test continue in execution, so that difference in the image does not hinder the execution, which can be tiresome in text fixing.
Lastly, going and viewing all images one by one is too lengthy a process. I would welcome a page that would show a test headlines and image summaries on one place, which I could browse and validate at once. Possibly with something like 'update all image differences' button.

**What did I achieve?**
Warnings are (if I were searching correctly) not yet a thing in Cypress, though I saw some issues on this topic. 

I make the tests continue with events
````
Cypress.on('fail', (error, runnable) => {
    if (error.message.includes("Image diff factor")) {
        console.warn("Test Regression Failure!", runnable.title, "\n", error.message);
        return false;
    }
    throw error;
});
````
which displays the error, 
![image](https://github.com/FRSOURCE/cypress-plugin-visual-regression-diff/assets/34658867/67a14ced-3cd0-4bad-a2c6-faf28967932c)

but does not issue any warning so that one needs to expand all test bodies and inspect them :/ It is also not very robust since it relies on the message contents, which could be improved in some sort of feature.

**Comment by FRSgit** [comment]
Unfortunately I believe that there is no good solution to your problem. As you said - [warnings in Cypress are not yet a thing](https://github.com/cypress-io/cypress/issues/4917), so that's off the table. The only other way to show messages would be in the terminal - but I think it's too easy to be omitted by the end user.

So in the end - I think that your solution is the best way of handling this issue.

**Comment by FRSgit** [comment]
Hey @Aiosa, I know it's been a while, but I wanted to circle back because two years ago I told you there was no good way to do this, and that's no longer true.

Yesterday I released v4.2.0 with something called Batch Review Mode, and honestly your post here was one of the things that pushed me to build it. Here's what it does once you it turn on via `pluginVisualRegressionBatchReviewMode: true`:

- `cy.matchImage()` stops throwing when a diff goes over the threshold. The test just carries on. Instead, a little floating button in the runner gets a counter badge that ticks up with every failure, so you can tell at a glance that something changed without digging through every test body.
- When the spec is done, the plugin throws one error saying how many snapshots failed. So the run is still red, nothing gets swallowed silently, and you don't need that separate "strict" re-run you mentioned.
- Click the button and you get a carousel with all the failing snapshots: test title, new vs. old side by side, a diff view, and two buttons per image, "Replace image" or "Skip this change". That's pretty much the review page you were asking for.
- There's also a "Show passing" checkbox. It lists images that changed but stayed under `maxDiffThreshold`, which is handy for cleaning up small drift that piles up over time.

So the `Cypress.on('fail')` hack that greps the error message can go. I'm also planning to make batch mode the default in v5.

One thing I didn't do is a "replace all" button. Approving is per image on purpose, mostly because I didn't want to make it too easy to accept a genuine regression by accident. If you're in the headless "just take everything from this deployment as the new baseline" situation, `updateImages: 'failures-only'` (from v4.0.0) does that: it rewrites only the baselines that failed and leaves the rest alone. If you still think a bulk approve in the UI would earn its place, open an issue and let's talk about it.

Proper Cypress warnings are still stuck upstream (cypress-io/cypress#4917), so badge plus one error at the end of the spec is as close as I can get for now.

If you give it a try, I'd genuinely like to hear how it fits your workflow. It's brand new and real-world feedback is what shapes OSS libraries like this one 🙂
